benches/metrics_bench_util/mod.rs

```diff
@@ -0,0 +1,292 @@
+                                out_addr.to_string(),
+                            ),
+                        );
+
+                        let mut rt = runtime();
+                        let (output_lines, topology) = rt.block_on(async move {
+                            let output_lines = CountReceiver::receive_lines(out_addr);
+                            let (topology, _crash) =
+                                start_topology(config.build().unwrap(), false).await;
+                            wait_for_tcp(in_addr).await;
+                            (output_lines, topology)
+                        });
+
+                        (input_lines, rt, topology, output_lines)
+                    },
+                    |(input_lines, mut rt, topology, output_lines)| {
+                        rt.block_on(async move {
+                            let sends = stream::iter(input_lines)
+                                .map(|lines| send_lines(in_addr, lines))
+                                .collect::<Vec<_>>()
+                                .await;
+                            future::try_join_all(sends).await.unwrap();
+
+                            topology.stop().compat().await.unwrap();
+
+                            let output_lines = output_lines.await;
+
+                            debug_assert_eq!(num_lines * num_writers, output_lines.len());
+
+                            output_lines
+                        });
+                    },
+                    BatchSize::PerIteration,
+                );
+            },
+        );
+    }
+
+    group.finish();
+}
+
+/// Here we perform some microbenchmarks on the metrics.
+#[inline]
+fn bench_micro(c: &mut Criterion, bench_name: &'static str) {
+    let mut group = c.benchmark_group(format!("{}/{}", bench_name, "micro"));
+    group.bench_function("bare_counter", |b| {
+        b.iter(|| {
+            counter!("test", 1);
+        });
+    });
+    group.bench_function("bare_counter_with_static_labels", |b| {
+        b.iter(|| {
+            counter!("test", 1, "my key" => "my value");
+        });
+    });
+    group.bench_function("bare_counter_with_dynamic_labels", |b| {
+        b.iter_batched(
+            || "my value".to_owned(),
+            |my_value| {
+                counter!("test", 1, "my key" => my_value);
+            },
+            BatchSize::SmallInput,
+        );
+    });
+    // A span that's not even entered.
+    group.bench_function("ununsed_span", |b| {
+        b.iter_batched_ref(
+            || span!(Level::ERROR, "my span"),
+            |_span| {
+                counter!("test", 1);
+            },
+            BatchSize::SmallInput,
+        );
+    });
+    // A span that's entered but without a counter invocation.
+    group.bench_function("span_enter_without_counter", |b| {
+        b.iter_batched_ref(
+            || span!(Level::ERROR, "my span"),
+            |span| {
+                let _enter = span.enter();
```

Comment:
```suggestion
                span.enter()
```

This might not be true in this particular case because I imagine `span.enter()` has side-effects, but sometimes the compiler will optimize code like this out since the value is unused. Returning it causes criterion to wrap it in `blackbox()` to ensure the compiler doesn't elide it.

Anyway, my eye is trained to look for not returning values from the iteration so even if it is unnecessary, it'd help not question the bench when when scanning :smile: Alternatively we can call `criterion::black_box()` ourselves.

This comment applies to the below benches too.

Reply:
Yeah, it's a good idea, just to be sure. Thanks for pointing this out!

Reply:
Done at 1704d5bdb

Reply:
Turns out we can't do it:

```
error: lifetime may not live long enough
   --> benches/metrics_bench_util/mod.rs:216:20
    |
216 |             |span| span.enter(),
    |              ----- ^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |              |   |
    |              |   return type of closure is Entered<'2>
    |              has type `&'1 mut Span`

error: lifetime may not live long enough
   --> benches/metrics_bench_util/mod.rs:226:17
    |
223 |             |span| {
    |              ----- return type of closure is Entered<'2>
    |              |
    |              has type `&'1 mut Span`
...
226 |                 enter
    |                 ^^^^^ returning this value requires that `'1` must outlive `'2`

error: lifetime may not live long enough
   --> benches/metrics_bench_util/mod.rs:237:17
    |
234 |             |span| {
    |              ----- return type of closure is Entered<'2>
    |              |
    |              has type `&'1 mut Span`
...
237 |                 enter
    |                 ^^^^^ returning this value requires that `'1` must outlive `'2`

error: lifetime may not live long enough
   --> benches/metrics_bench_util/mod.rs:255:17
    |
252 |             |span| {
    |              ----- return type of closure is Entered<'2>
    |              |
    |              has type `&'1 mut Span`
...
255 |                 enter
    |                 ^^^^^ returning this value requires that `'1` must outlive `'2`

error: lifetime may not live long enough
   --> benches/metrics_bench_util/mod.rs:269:17
    |
266 |             |span| {
    |              ----- return type of closure is Entered<'2>
    |              |
    |              has type `&'1 mut Span`
...
269 |                 enter
    |                 ^^^^^ returning this value requires that `'1` must outlive `'2`

error: lifetime may not live long enough
   --> benches/metrics_bench_util/mod.rs:289:17
    |
286 |             |span| {
    |              ----- return type of closure is Entered<'2>
    |              |
    |              has type `&'1 mut Span`
...
289 |                 enter
    |                 ^^^^^ returning this value requires that `'1` must outlive `'2`

error: aborting due to 6 previous errors
```